service/tablet_allocator.cc

```diff
@@ -1021,21 +1000,60 @@ class load_balancer {
                 if (skip) {
                     lblogger.debug("Replication constraint check failed, unable to emit migration for replica ({}, {}) to co-habit the replica ({}, {})",
                         t2_id, src, t1_id, dst);
-                    return make_ready_future<>();
+                    return;
                 }
 
                 auto mig = create_migration_info(t2_id, src, dst);
-                auto mig_streaming_info = get_migration_streaming_info(_tm->get_topology(), *t2.info, mig);
+                auto mig_streaming_info = get_migration_streaming_info(_tm->get_topology(), *t2.desc.info, mig);
                 if (!can_accept_load(nodes, mig_streaming_info)) {
                     // FIXME: we can try another pair of non-colocated replicas of same sibling tablets.
                     lblogger.debug("Load limit reached, unable to emit migration for replica ({}, {}) to co-habit the replica ({}, {})",
                         t2_id, src, t1_id, dst);
-                    return make_ready_future<>();
+                    return;
                 }
                 apply_load(nodes, mig_streaming_info);
 
                 lblogger.info("Created migration for replica ({}, {}) to co-habit same shard as ({}, {})", t2_id, src, t1_id, dst);
                 plan.add(std::move(mig));
+            };
+
+            co_await tmap.for_each_sibling_tablets([&] (tablet_desc t1, std::optional<tablet_desc> t2_opt) -> future<> {
+                // Be optimistic about migrating tablets, as if they succeeded.
+                // Merge finalization will have to recheck that all sibling tablets are co-located.
+
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
+                auto t2 = *t2_opt;
+
+                auto r1 = get_replicas(t1);
```

Comment:
Are those per dc?  I don't see that filtering in `get_replicas`...
Seems like a pre-existing issue that needs to be fixed (i.e. the current implementation is not fully dc-aware)

Reply:
Yes, it is. Co-location plan is called in the context of a given DC, and nodes used for filtering in replicas (in get_replicas()) only belong to the DC being worked on.

Reply:
I think the name get_replicas() is misleading, I will rename it to get_sorted_dc_replicas() or something.

Reply:
Why "sorted" by the way?
What is it sorted on?
Isn't it just ordered?

Reply:
> Yes, it is. Co-location plan is called in the context of a given DC, and nodes used for filtering in replicas (in get_replicas()) only belong to the DC being worked on.

Got it, thanks.

Reply:
> Why "sorted" by the way?
> What is it sorted on?
> Isn't it just ordered?

It's sorted by host id to handle the scenario t1: [n1,n2], t2:[n2,n1], both of which belonging to same rack. Which is why we'll only be able to enable table with views on RF=#racks to keep consistency

Reply:
I see, thanks.